simularium/simularium-website
# Feature/color picker

Problem
=======
Closes [#294](https://github.com/simularium/simularium-website/issues/294) - Users want to choose colors for their agents.

Solution
========
Agents can be selected by clicking on the color swatches in the side panel and interacting with a color picker component. Will allow users to change colors for agents and subagents, and keep track of recent color selections.

**Redux:** Actions, types, constants, reducer functions and initial state are all updated to handle setting `ColorChange` and `RecentColors`. As with most changes to redux, this touches a bunch of files but is pretty straightforward. 

**Selectors:** Updated in selection state branch, and ViewerPanel, to account for new typing of `SelectionStateInfo` that includes `ColorChange`.

**UI Components:** ColorPicker is new. We are implementing the react-colorful library to handle some of the functionality and then styling their components inside an antd popover.  There is a fair amount of CSS included some `!important` tags and various nested antd classes. Will happly remove/clean up rules if I missed opportunities to simplify. 

Changes to CheckBox and CheckBox tree are for passing down props to ColorPicker and resizing the swatches and containers to allow click behavior. Additionally, we had to swap the order of the `star, checkbox, color` to `star, color swatch, checkbox`, which may seem trivial, but because they're shared checkboxes they're being rendered as groups. 

**Note:** confirmed with UX team that shifting hue selection bar from vertical to horizontal orientation was acceptable.

with @meganrm

## Type of change
* New feature (non-breaking change which adds functionality)

Steps to Verify:
----------------

- npm run start
- select agents and sub agents, confirm that the colors change in the viewer and in the side panel
- if sub agents are different colors, the parent color swatch should be blank with a white border
- recently selected colors should display in the color picker when selecting different agents

<img width="310" alt="Screenshot 2023-12-07 at 4 31 06 PM" src="https://github.com/simularium/simularium-website/assets/5170636/e6687eb8-c692-4242-8ee4-dce128fe9466">
<img width="303" alt="Screenshot 2023-12-07 at 4 31 14 PM" src="https://github.com/simularium/simularium-website/assets/5170636/86bc2105-7a0a-455f-96fb-79a8cae06cb7">
